### PR TITLE
extract library

### PR DESCRIPTION
**Pull request by MarcoIeni**
A step towards #67 
To keep things easier I just created a new lib.rs
In the future, we could extract this lib.rs file into a `core` crate, that doesn't depend on clap.

### CODE REVIEW

File: src/lib.rs
Comment (MarcoIeni) [inline src/lib.rs:41]: I tried to group conflicting cli arguments together in enums.

File: src/lib.rs
Comment (MarcoIeni) [inline src/lib.rs:50]: I saw that this rustdoc also requires the Baseline rustdoc.
At the moment this is not captured by the type system. How would you deal with this?

File: src/lib.rs
Comment (MarcoIeni) [inline src/lib.rs:75]: this function is very similar to [partition_packages](https://docs.rs/clap-cargo/latest/clap_cargo/struct.Workspace.html#method.partition_packages).
Initially, I wanted to use it, but since the struct `clap_cargo::Workspace` is non-exhaustive, it cannot be constructed.

File: src/lib.rs
Comment (MarcoIeni) [inline src/lib.rs:120]: builder pattern: https://rust-lang.github.io/api-guidelines/type-safety.html#non-consuming-builders-preferred

File: src/lib.rs
Comment (MarcoIeni) [inline src/lib.rs:172]: I'm not sure about this.

File: src/lib.rs
Comment (MarcoIeni) [inline src/lib.rs:304]: the report should contain the failing checks.
For now, the library prints to stdout and just reports if it fails or succeeds.

File: src/config.rs
Comment (MarcoIeni) [inline src/config.rs:18]: I added this to fix a clippy lint

**Review by obi1kenobi — COMMENTED** [review]
Very excited about this, thanks for putting it together!

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs:51]: In this way, I don't have to mess with lifetimes :)

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs:280]: Shall we merge this pr after #270?

Comment (tonowak) [reply src/lib.rs:282]: #270 has been merged.

Comment (MarcoIeni) [reply src/lib.rs]: merged here in [7712fae](https://github.com/obi1kenobi/cargo-semver-checks/pull/314/commits/7712fae0eb1c07dbe7c77e6d699f06b5c810ad02)

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs:61]: So we want to create a RustdocSource struct that can be used both by the Baseline and Current rustdoc?

> already generated, just load it from an existing file

```rust
enum RustdocSource {
    RustDoc(PathBuf)
}
```

> generate it based on code at some path, optionally first checking out a gitrev (maybe better split into two cases? what do you think?)



```rust
enum RustdocSource {
    RustDoc(PathBuf),
    Root(PathBuf),
    // (path to root, rev)
    Revision(PathBuf, String)
}
```

> generate it based on a registry version

```rust
enum RustdocSource {
    RustDoc(PathBuf),
    Root(PathBuf),
    // (path to root, rev)
    Revision(PathBuf, String),
    Version(Option<String>),
}
```

This looks similar to the current Baseline enum. Is it valid for `Current`, too?
If I didn't get it, can you give me the structure of the enum you were thinking about?

Comment (obi1kenobi) [reply src/lib.rs:61]: I think it looks reasonable. It currently isn't valid for `Current` in the CLI, but I think it should be in the lib. Not much reason not to: that would make the types more complex while providing less functionality.

Small nit: let's call the variant `Rustdoc` instead, since the underlying tool doesn't capitalize the `D` either.

**Comment by MarcoIeni** [comment]
Hey @tonowak, if I'm blocking you, feel free to contribute to this pr, yes 👍 
For example, today I won't have time to look into [this](https://github.com/obi1kenobi/cargo-semver-checks/pull/314#discussion_r1083390720)

File: src/lib.rs
Comment (MarcoIeni) [inline src/lib.rs:85]: probably for this pr this is good enough.
As a future improvement, maybe we could use a temporary directory as "target".

Comment (obi1kenobi) [reply src/lib.rs:85]: I'm quite confused about what this function does, and the name isn't helping me out much. Is there a more descriptive name we could use for it?

I'm also confused by the fact that the function only works for the `Root` and `Revision` variants, but generates a `RegistryBaseline` result on success.

Comment (MarcoIeni) [reply src/lib.rs]: I inlined the function. In this way, it should be easier to review because it is easier to compare with the [original code](https://github.com/obi1kenobi/cargo-semver-checks/blob/ee1fa79054ccaaab762845f920bd95782e246fff/src/main.rs#L117). 👍

**Comment by MarcoIeni** [comment]
> Feel free to commit directly into the branch, or to raise prs against it.

Nevermind Tomasz, I found some time to work on this tonight 👍

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs]: In 1190770 I have made it impossible for this to happen

**Comment by tonowak** [comment]
No problem! I see you're progressing faster than I can help :)

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:26]: We should probably `#[derive(Debug)]` and `#[non_exhaustive]` at least for all these pub structs and enums, unless there's a reason not to?

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:31]: Is this different than `GlobalConfig.log_level`? I'm confused by the duplication.

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:46]: Just breaking a long line.
```suggestion
    /// Generate the rustdoc file from the project root directory,
    /// i.e. the directory containing the crate source.
```

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:48]: Is Rust doc comment style to include or omit the parens when naming functions? I genuinely am not sure, we should probably mirror what `std` does here.

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:74]: We probably want a doc comment here as well.

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:82]: Just breaking another long line. Generally would be nice to keep line length below 100 chars max.
```suggestion
    /// Path to the Rustdoc json file.
    /// Use this option when you have already generated the rustdoc file.
```

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:81]:
```suggestion
#[derive(Debug)]
enum RustdocSource {
```
for all types unless there's a reason not to?

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:254]: Hmm is `--current-rustdoc` not supported together with `--baseline-version`? I'm not sure why that might be the case? I'd have expected it to just assume that the manifest is in the current directory and up-to-date in that case, and bail if it is not.

If indeed it doesn't work today, then I will update my own prior — no code change needed in that case :)

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:339]: I believe these are currently unimplemented in the CLI, so I'm fine with leaving a panic in this branch.

But let's make it `unimplemented!()` with a suitable message instead of a bare `todo!()`, which makes it seem like we forgot to add it into the PR rather than making an explicit choice to not include it for now.

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:43]:
```suggestion
#[non_exhaustive]
pub struct SemverQuery {
```
We may want to add new fields in the future here.

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:349]:
```suggestion
#[non_exhaustive]
#[derive(Debug)
pub struct Report {
```
I know `#[non_exhaustive]` is redundant here at the moment, but better safe than sorry. We don't want to end up in a situation where we've accidentally removed the last private field without realizing, and we're then locked into a constructible exhaustive struct until the next major version.

One day `cargo-semver-checks` will warn about "semver-regrettable" situations like that.

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:136]:
```suggestion
#[non_exhaustive]
#[derive(Debug, Default, PartialEq, Eq, Clone)]
pub enum ScopeSelection {
```
Unless there's a reason not to?

**Review by obi1kenobi — COMMENTED** [review]
Sorry for the delay! I'm pretty happy with the shape of things here. Except for that one block that I'd love to get @epage's eyes on, I think this is looking pretty good. Just some minor "polish-level" suggestions that should be quickly resolvable.

Excited to ship this soon!

File: src/lib.rs
Comment (tonowak) [inline src/lib.rs:306]: One of the `selected.publish == ...` is not needed.

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs:26]: Do you want to add `#[non_exhaustive]` even if the struct fields are private?
I don't think it has any effect in this case.

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs]: done in https://github.com/obi1kenobi/cargo-semver-checks/pull/314/commits/f626d7ebaed9becbf30edaaf24ff273a75fa0830

File: src/query.rs
Comment (MarcoIeni) [reply src/query.rs:44]: done in https://github.com/obi1kenobi/cargo-semver-checks/pull/314/commits/368ee5b0c0e866e8dedc00d079ff4ac63cd764e1

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs:80]: done in [5a334ba](https://github.com/obi1kenobi/cargo-semver-checks/pull/314/commits/5a334bac2269731663c2c0f377911b61a1bc8b83)

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs]: By looking at [this](https://doc.rust-lang.org/std/env/index.html) page, it seems that the std uses the () convention.
I changed this in a496d420ecd96a1fe338bdc96a92f8960350a7a5

File: src/lib.rs
Comment (tonowak) [reply src/lib.rs]: > It currently isn't valid for Current in the CLI, but I think it should be in the lib. 

FYI @MarcoIeni, the recent merged chain of PRs (which ended with https://github.com/obi1kenobi/cargo-semver-checks/pull/351) made it so that the `current` crate's rustdoc is generated in the same way as the `baseline`'s rustdoc, so now it's possible to create the `current` rustdoc from registry and from git revision -- your `RustdocSource` now is also fully valid for the `current` crates.

Comment (tonowak) [reply src/lib.rs]: And also, this chain of PRs renamed quite a lot of things in the source code, so there might be a big merge conflict, sorry :( I can try to merge main into this branch when I'll have some free time.

Comment (obi1kenobi) [reply src/lib.rs]: If you could help with the merge, that would be great @tonowak. Probably the highest impact next thing you can do right now, TBH.

Comment (tonowak) [reply src/lib.rs]: Ok, I'll do it in max few days.

**Comment by tonowak** [comment]
I've merged `main` into this branch. @obi1kenobi can you *carefully* review the merge? I had to change some logic in it.

File: tests/verify_binary_contains_lints.rs
Comment (tonowak) [inline tests/verify_binary_contains_lints.rs:8]: I've made this change temporarily so that the test passes locally, we should revert this change.

From what I understand, the flag `--baseline-root` was meant to take a directory to the project, but the old code somehow worked both when it received a directory and when it received a path to a manifest (and I believe that we didn't even know what we had this feature). Passing a directory is "more correct" in the sense that it aligns with the description from `--help`, but I think there might be some users who might rely on the accidental feature of passing manifests (like this test), so I think that we should make the current code also work for both directories and manifests.

File: src/lib.rs
Comment (tonowak) [inline src/lib.rs:439]: As the comment says, I'll come back to this function in a few days.

File: src/lib.rs
Comment (tonowak) [inline src/lib.rs:248]: Now that I think about this -- the logic behind this function is not obvious at all, I should document this.

File: tests/verify_binary_contains_lints.rs
Comment (obi1kenobi) [reply tests/verify_binary_contains_lints.rs:8]: I also think it's reasonable to be able to extrapolate the directory if we're passed the manifest. Seems like an easy way to do the right thing even if the user did the wrong thing.

Nice catch! We got lucky this time, so let's add a test so we don't have to rely on luck for next time?

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs:28]: done! :)

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs:88]: done :)

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs]: After the changes from [tonowak](https://github.com/tonowak) the `todo` is not there anymore 👍

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs]: fixed in [371ccab](https://github.com/obi1kenobi/cargo-semver-checks/pull/314/commits/371ccabd22e9253d6aa0ef85edd7774fbe327f42)

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs]: After the changes from [tonowak](https://github.com/tonowak), the `anyhow::bail` is not there anymore 👍

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs:33]: I decided not to expose GlobalConfig to the user. I'm only exposing the log level.
That's because if the user wants to use `cargo-semver-checks` as a library, I imagine `cargo-semver-checks` won't write to stdout or stderr in the future. So it doesn't make sense to allow the user to customize stdout and stderr.

**Comment by MarcoIeni** [comment]
Thanks for helping Tomasz 😄 
I think I have addressed all the PR comments. I won't edit this PR anymore unless you still need something else from my side 👍

File: tests/verify_binary_contains_lints.rs
Comment (tonowak) [reply tests/verify_binary_contains_lints.rs:8]: OK, I've implemented it and I've added some tests. Also, the previous error messages of passing directory to `--manifest-path` were not clear. I'm not sure whether I made the right choice, but I've made it so it's possible to pass both a manifest and a directory path to the `--manifest-path` flag.

File: src/lib.rs
Comment (tonowak) [reply src/lib.rs]: Done in 2449acd377ed1792127a10c0ffed211094664d1b.

File: src/lib.rs
Comment (tonowak) [reply src/lib.rs]: We discussed on Zulip whether this is the best idea and we've decided to instead rely on the XDG cache directory: 2449acd377ed1792127a10c0ffed211094664d1b

File: src/lib.rs
Comment (tonowak) [reply src/lib.rs]: Done.

File: src/lib.rs
Comment (tonowak) [inline src/lib.rs:101]: Maybe it would be better to call it `Registry` instead of `Version`? It would make it more clear that it requires internet access to use it and from where this specific version is being taken from.

Comment (MarcoIeni) [reply src/lib.rs:101]: What about `RegistryVersion`? 😄 
The user when seeing `Registry` with a string might think about `Registry("crates.io")` for example

Comment (MarcoIeni) [reply src/lib.rs:101]: Naming is hard 😁
`VersionFromRegistry` is more verbose, but more clear maybe. 🤔

Comment (obi1kenobi) [reply src/lib.rs:101]: I'm fine with either `RegistryVersion` or `VersionFromRegistry`. I'm not that worried about people thinking that they need to pass a version of a registry, but I'm willing to accept I may be overly optimistic if you think the other option is better.

Comment (tonowak) [reply src/lib.rs]: What do you both think about 8485cc5dfe51a91ba2e8fe2b071c631f24f67cfb?

File: src/lib.rs
Comment (tonowak) [reply src/lib.rs:304]: In #207 I've made a temporary fix 7ecf5873242bc0defbb1eb739876cdc778614161 so that I can progress further. In the case where this fix is good enough, I can move it here, otherwise we should discuss better alternatives.

**Comment by tonowak** [comment]
The CLI needs some polishing, but I'm not experienced in stuff like that -- @obi1kenobi can you review it?

I'm thinking that `semver-crater` might be useful not only for data gathering, but also for convincing people that the tool is worthy. The `semver-crater` allows to easily check whether there has been any issues in the user's crate (just run a single `cargo` command that install `cargo-semver-checks` and runs it on all releases of the user's crate) and to show people "look, some popular crates (like clap) had issues, here is how in a single command you can check that it is indeed true and the tool even detects it".

Assuming `semver-crater` would be used in this manner, we have to think about how to make its CLI convenient to use.

File: src/lib.rs
Comment (MarcoIeni) [reply src/lib.rs]: I think it's fine :)
I would not try to make the library perfect in this PR. Otherwise we will have to deal with too many merge conflicts.

I guess (and hope) that the library will be extracted in a separate crate, so we can still change the library API in future PRs.

Comment (obi1kenobi) [reply src/lib.rs]: This PR is on my to-do list to review today. Sorry for the delay! Trustfall ended up in the spotlight (it was [GitHub's #​1 most-starred project](https://twitter.com/PredragGruevski/status/1623711641127747584) a couple of days ago) and I've been super busy over there for the last couple of days.

I know I probably just merged another PR that causes another merge conflict, and I'm happy to fix the mess of my own making 😅

File: src/lib.rs
Comment (obi1kenobi) [reply src/lib.rs:304]: In https://github.com/obi1kenobi/cargo-semver-checks/commit/7ecf5873242bc0defbb1eb739876cdc778614161 it seems like there's still an `unknown` branch. That looks like it could work if `current` is given by gitrev, but won't work with a registry version nor with a premade rustdoc file.

Can we add an explicit `panic!()` in the cases we know for sure won't work, instead of setting a bogus value and then breaking in a confusing way?

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:252]: I suspect this won't work on Windows, unfortunately. Let's instead use [the `directories` crate](https://crates.io/crates/directories) which is the known-working cross-platform way to do this:
```rust
if let Some(proj_dirs) = ProjectDirs::from("cargo-semver-checks") {
    proj_dirs.config_dir();
    // probable dirs (still need to be created! the library won't make them for us)
    // Lin: /home/alice/.config/cargo-semver-checks
    // Win: C:\Users\Alice\AppData\Roaming\cargo-semver-checks\config
    // Mac: /Users/Alice/Library/Application Support/cargo-semver-checks
}
```

**Review by obi1kenobi — COMMENTED** [review]
This looks in pretty good shape. Let's target merging tomorrow or Monday morning if possible, so I can include this in the next release (also planned for Monday at the moment).

**Comment by obi1kenobi** [comment]
I took care of the merge, it wasn't that bad because I reviewed the code that caused a conflict. Lmk what you think about the `panic!()` branch and using the cross-platform-compatible app dir crate!

Really looking forward to merging this soon! 🚀🎉

**Comment by MarcoIeni** [comment]
> Lmk what you think about the panic!() branch and using the cross-platform-compatible app dir crate!

Is this question for me? If yes, what's the panic!() branch?

**Comment by obi1kenobi** [comment]
The two remaining items needed to merge this are:
- [This `<unknown>` branch](https://github.com/obi1kenobi/cargo-semver-checks/pull/314/files#r1103730310) that seems likely to cause an obscure error if someone hits it, and I'm recommending an explicit `panic!()` instead. It's based on changes @tonowak made.
- [Using the cross-platform `directories` crate instead of `xdg`](https://github.com/obi1kenobi/cargo-semver-checks/pull/314#discussion_r1103730464) so that our non-Unix users aren't left behind.

I unfortunately have had to shift my primary focus to Trustfall for a bit, given [the recent massive uptick in interest in the project](https://twitter.com/StarHistoryHQ/status/1625421205321793536). I think the remaining items aren't much work and I trust both of you to do a good job on them.

**Comment by MarcoIeni** [comment]
I replaced `xdg` with `directories`. I leave the other fix to tonowak :)

**Comment by tonowak** [comment]
Thanks :) I'll probably do the other fix in the next 48h.

**Review by obi1kenobi — APPROVED** [review]
This looks polished enough to release, let's not block @MarcoIeni's `release-plz` work any longer. We will certainly keep improving the library APIs over time, we don't have to get to "perfect" on day 1.

Thanks, everyone!

**Comment by MarcoIeni** [comment]
Thanks!
This week I'll work on extracting the library into a separate crate.
In this way, the users of the library don't need to compile `clap` and other specific libraries used only in the cli.

Some naming ideas:

- `semver-checks`
- `cargo-semver-checks-core`

Please, reserve in crates.io the one you prefer :)

**Comment by obi1kenobi** [comment]
I grabbed both, just to be extra safe. `semver-checks` seems good to me, but I don't have a strong preference.

My 2 cents: splitting the library into a separate crate would definitely be nice, but it doesn't seem that critical to me. The community impact of having semver-checking made automatic via `release-plz` seems bigger, and we can always split the library into a separate crate and cut down compile times later.

Of course, `release-plz` is your crate, this is your work, and it's your call to make :)

**Comment by tonowak** [comment]
Maybe it would be reasonable to have the `clippy` and CLI under a default flag?

**Comment by MarcoIeni** [comment]
> CLI under a default flag

This would prevent us to ship breaking changes independently for the cli and for the library.

For example, if we change a command line argument, should it be a breaking change?

Having two separate crates allows us to ship breaking changes for cli and the lib independenlty

**Pull request merged**
